### PR TITLE
tmt/convert.py: add typing

### PR DESCRIPTION
**Pull request by ukulekek**

### CODE REVIEW

File: tmt/convert.py [hunk 1]
```diff
@@ -8,10 +8,15 @@
 import re
 import subprocess
 from io import open
+from typing import Any, Dict, List, Optional, Tuple, Union, cast
 
 import fmf.utils
+# Used only for typing
+import nitrate as _nitrate
```
Comment (happz) [inline tmt/convert.py:15]: Why the leading underscore? I suppose there's a reason for it, but it's not obvious, please add a comment.

File: tmt/convert.py [hunk 2]
```diff
@@ -116,7 +125,7 @@ def read_manual_data(testcase):
     return md_content
 
 
-def html_to_markdown(html):
+def html_to_markdown(html: str) -> Any:
```
Comment (happz) [inline tmt/convert.py:128]: Isn't `md_handler.handle()` returning a string? It seems to me `str` would be better here, I'm pretty sure return value of `html_to_markdown` is never going to be `bool` or `int` or `List[Tuple[Dict[str, bool]]]`, we can make it stricter.

File: tmt/convert.py [hunk 3]
```diff
@@ -201,23 +218,27 @@ def read_datafile(path, filename, datafile, types, testinfo=None):
         testinfo = datafile
     # Beaker task name
     try:
-        beaker_task = re.search(regex_task, testinfo).group(1)
+        search_result = re.search(regex_task, testinfo)
+        assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:222]: Are we absolutely sure `search_result` is going to be a valid value? What if `filename` is `Makefile` but said Makefile is missing `Name: ...` line because the user removed it? We'd drop a nasty assert traceback on the user's head when we could instead raise a nice `GeneralError` saying "sorry, could find line matching this pattern, are you sure it's there?"

File: tmt/convert.py [hunk 4]
```diff
@@ -201,23 +218,27 @@ def read_datafile(path, filename, datafile, types, testinfo=None):
...
         data['extra-task'] = beaker_task
         data['extra-summary'] = beaker_task
     except AttributeError:
         raise ConvertError("Unable to parse 'Name' from testinfo.desc.")
     # Summary
     try:
-        data['summary'] = re.search(
-            regex_summary, testinfo, re.M).group(1)
+        search_result = re.search(regex_summary, testinfo, re.M)
+        assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:232]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 5]
```diff
@@ -201,23 +218,27 @@ def read_datafile(path, filename, datafile, types, testinfo=None):
...
+        data['summary'] = search_result.group(1)
         echo(style('summary: ', fg='green') + data['summary'])
     except AttributeError:
         pass
     # Test script
     try:
-        data['test'] = re.search(
-            regex_test, datafile, re.M).group(1)
+        search_result = re.search(regex_test, datafile, re.M)
+        assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:240]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 6]
```diff
@@ -236,26 +257,29 @@ def read_datafile(path, filename, datafile, types, testinfo=None):
         raise ConvertError("Unable to open '{0}'.".format(test_path))
     # Contact
     try:
-        data['contact'] = re.search(
-            regex_contact, testinfo, re.M).group(1)
+        search_result = re.search(regex_contact, testinfo, re.M)
+        assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:261]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 7]
```diff
@@ -236,26 +257,29 @@ def read_datafile(path, filename, datafile, types, testinfo=None):
...
     except AttributeError:
         pass
 
     if filename == 'Makefile':
         # Component
         try:
-            data['component'] = re.search(
-                r'^RunFor:\s*(.*)', testinfo, re.M).group(1).split()
+            search_result = re.search(r'^RunFor:\s*(.*)', testinfo, re.M)
+            assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:271]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 8]
```diff
@@ -236,26 +257,29 @@ def read_datafile(path, filename, datafile, types, testinfo=None):
...
                  ' '.join(data['component']))
         except AttributeError:
             pass
 
     # Duration
     try:
-        data['duration'] = re.search(
-            regex_duration, testinfo, re.M).group(1)
+        search_result = re.search(regex_duration, testinfo, re.M)
+        assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:281]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 9]
```diff
@@ -291,8 +315,9 @@ def read_datafile(path, filename, datafile, types, testinfo=None):
     if filename == 'Makefile':
         # Convert Type into tags
         try:
-            makefile_type = re.search(
-                r'^Type:\s*(.*)', testinfo, re.M).group(1)
+            search_result = re.search(r'^Type:\s*(.*)', testinfo, re.M)
+            assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:319]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 10]
```diff
@@ -438,11 +472,13 @@ def read(
...
 
         # Warn if makefile has extra lines in run and build targets
-        def target_content(target):
+        def target_content(target: str) -> List[str]:
             """ Extract lines from the target content """
             regexp = rf"^{target}:.*\n((?:\t[^\n]*\n?)*)"
             try:
-                target = re.search(regexp, datafile, re.M).group(1)
+                search_result = re.search(regexp, datafile, re.M)
+                assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:480]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 11]
```diff
@@ -653,12 +700,13 @@ def read_nitrate_case(testcase, makefile_data=None, general=False):
...
         else:
             try:
+                assert makefile_data is not None
                 # Use contact from Makefile if it's there and email matches
                 if re.search(testcase.tester.email, makefile_data['contact']):
                     data['contact'] = makefile_data['contact']
                 else:
                     raise KeyError
-            except (KeyError, TypeError):
+            except (KeyError, TypeError, AssertionError):
```
Comment (happz) [inline tmt/convert.py:709]: Nope, never. `assert` is not executed when running Python with `-O` option, which means this exception would never be raised. `assert` helps mypy understand what's happening, but don't rely on it when it comes to the function of the code. You may use different forms of checks if `makefile_data == None` needs to be sanitized somehow.

File: tmt/convert.py [hunk 12]
```diff
@@ -736,8 +784,9 @@ def read_nitrate_case(testcase, makefile_data=None, general=False):
 
     # Extend bugs detected from Makefile with those linked in Nitrate
     try:
+        assert makefile_data is not None
         data['link'] = makefile_data['link'].copy()
-    except (KeyError, TypeError):
+    except (KeyError, TypeError, AssertionError):
```
Comment (happz) [inline tmt/convert.py:789]: Same comment WRT use of `AssertionError`.

File: tmt/convert.py [hunk 13]
```diff
@@ -846,7 +896,9 @@ def relevancy_to_adjust(relevancy):
     for line in re.split(r'\s*\n\s*', relevancy.strip()):
         # Extract possible comments
         try:
-            line, rule['because'] = re.search(RELEVANCY_COMMENT, line).groups()
+            search_result = re.search(RELEVANCY_COMMENT, line)
+            assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:900]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 14]
```diff
@@ -856,7 +908,9 @@ def relevancy_to_adjust(relevancy):
 
         # Split rule
         try:
-            condition, decision = re.search(RELEVANCY_RULE, line).groups()
+            search_result = re.search(RELEVANCY_RULE, line)
+            assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:912]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 15]
```diff
@@ -875,8 +929,9 @@ def relevancy_to_adjust(relevancy):
         expressions = list()
         for expression in re.split(r'\s*&&?\s*', condition):
             try:
-                left, operator, right = re.match(
-                    RELEVANCY_EXPRESSION, expression).groups()
+                search_result = re.search(RELEVANCY_EXPRESSION, expression)
+                assert search_result is not None
```
Comment (happz) [inline tmt/convert.py:933]: Same comment I made towards `regex_task`

File: tmt/convert.py [hunk 16]
```diff
@@ -131,15 +140,16 @@ def html_to_markdown(html):
...
     """ Write gathered metadata in the markdown format """
     to_print = ""
     if content['setup']:
         to_print += "# Setup\n" + content['setup'] + '\n\n'
     if content['action'] or content['expected']:
         to_print += "# Test\n\n"
         if content['action']:
-            to_print += "## Step\n" + content['action'] + '\n\n'
+            to_print += "## Step\n" + \
+                content['action'] + '\n\ntemplates/Dockerfile.tailored.j2'
```
Comment (thrix) [inline tmt/convert.py:152]: @ukulekek ups :) this seems an unrelated change

File: tmt/convert.py [hunk 17]
```diff
@@ -171,13 +182,19 @@ def add_bug(bug, data, system=SYSTEM_BUGZILLA):
     echo(style('relates: ', fg='green') + new_link['relates'])
 
 
-def read_datafile(path, filename, datafile, types, testinfo=None):
+def read_datafile(path: str,
+                  filename: str,
+                  datafile: str,
+                  types: List[str],
+                  testinfo: Optional[str] = None) -> Tuple[str,
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/convert.py:189]: Blame autopep, it has really, really weird ideas about how code should be formatted... Black sucks, but compared to what autopep does, it's almost acceptable :/

File: tmt/convert.py [hunk 18]
```diff
@@ -516,7 +552,14 @@ def target_content(target):
     return common_data, individual_data
 
 
-def read_nitrate(beaker_task, common_data, disabled, general):
+def read_nitrate(beaker_task: str,
+                 common_data: Dict[str,
+                                   Any],
+                 disabled: bool,
+                 general: bool) -> Tuple[Dict[str,
```
Comment (thrix) [inline tmt/convert.py:559]: @ukulekek please create an alias for this crazy return type

File: tmt/convert.py [hunk 19]
```diff
@@ -635,9 +678,13 @@ def read_nitrate(beaker_task, common_data, disabled, general):
     return common_data, individual_data
 
 
-def read_nitrate_case(testcase, makefile_data=None, general=False):
+def read_nitrate_case(testcase: TestCase,
+                      makefile_data: Optional[Dict[str,
```
Comment (thrix) [inline tmt/convert.py:682]: I am thinkin about creating aliases also for `Dict[str, Any]` the code would become more readable

File: tmt/convert.py [hunk 20]
```diff
@@ -311,8 +336,16 @@ def read_datafile(path, filename, datafile, types, testinfo=None):
...
-def read(
-        path, makefile, restraint, nitrate, purpose, disabled, types, general):
+def read(path: str,
+         makefile: bool,
+         restraint: bool,
+         nitrate: bool,
+         purpose: bool,
+         disabled: bool,
+         types: List[str],
+         general: bool) -> Tuple[Dict[str,
```
Comment (thrix) [inline tmt/convert.py:346]: @ukulekek let's create an alias for thic crazy return type

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/convert.py (hunk 1 above)
Comment (ukulekek) [reply tmt/convert.py]: Wasn't needed after all, removed

File: tmt/convert.py (hunk 3 above)
Comment (ukulekek) [reply tmt/convert.py]: Thank you. I removed all asserts and added conditional instead. However, I removed a couple of `try except` constructions as they were not needed anymore

File: tmt/convert.py (hunk 18 above)
Comment (ukulekek) [reply tmt/convert.py]: Created

File: tmt/convert.py (hunk 19 above)
Comment (ukulekek) [reply tmt/convert.py]: Added, thanks

File: tmt/convert.py [hunk 21]
```diff
@@ -8,15 +8,24 @@
 import re
 import subprocess
 from io import open
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Union
 
 import fmf.utils
+# Used only for typing
 from click import echo, style
```
Comment (happz) [inline tmt/convert.py:15]: I don't understand the comment - maybe it's a leftover from `_nitrate`?

File: tmt/convert.py [hunk 22]
```diff
@@ -129,10 +142,11 @@ def html_to_markdown(html):
         markdown = ""
     else:
         markdown = md_handler.handle(html).strip()
+    assert isinstance(markdown, str)
```
Comment (happz) [inline tmt/convert.py:145]: I'd probably go with `cast()` here. Line 142 introduces the variable as a string, `cast(str, md_handler.handle(...)).strip()` would make it explicit that only string is assigned to `markdown`, and it'd be applied at the point where "string but it's not visible" value enters tmt code. `assert` is a solution when it's much harder to introduce the correct type, or when it's too late to annotate.

File: tmt/convert.py [hunk 23]
```diff
@@ -8,15 +8,24 @@
...
 from click import echo, style
 
+import tmt.export
 import tmt.utils
 from tmt.utils import ConvertError
 
 log = fmf.utils.Logging('tmt').logger
 
+# It is not possible to use TypedDict here, because all keys are unknown
+DataType = Dict[str, Any]
```
Comment (happz) [inline tmt/convert.py:24]: I'd propose a different name, there is way too many "data" packages flying around. Can we find more specific "data", e.g. `NitrateDataType` or `TCMSDataType`? What kind of data it'd contain?

File: tmt/convert.py (hunk 22 above)
Comment (ukulekek) [reply tmt/convert.py]: Thank you, it works

File: tmt/convert.py (hunk 23 above)
Comment (ukulekek) [reply tmt/convert.py]: Agree, renamed to `NitrateDataType`. Looks like it contains metadata which `Makefile` contained before in beakerlib tests

**Review by happz — APPROVED** [review]

File: tmt/convert.py [hunk 24]
```diff
@@ -389,6 +413,7 @@ def read(
             restraint_file = True
             echo(style('Restraint ', fg='blue'), nl=False)
```
Comment (thrix) [inline tmt/convert.py:414]: @ukulekek maybe it would be better to add an else condition here and raise some resonable error, instead of the assert below. The thing is, if somebody does not pass required variables to this function, I would rather like to see a nice error than a assertion error. Seems to me a clear and unhandled error condition. WDYT @ukulekek

File: tmt/convert.py [hunk 25]
```diff
@@ -652,7 +682,9 @@ def read_nitrate(beaker_task, common_data, disabled, general):
     return common_data, individual_data
 
 
-def extract_relevancy(notes, field):
+def extract_relevancy(
+        notes: str,
+        field: tmt.utils.StructuredField) -> Optional[Union[str, List[str]]]:
```
Comment (thrix) [inline tmt/convert.py:687]: @ukulekek could we add a type alias for `Union[str, List[str]]`, maybe call it `RelevancyType` ?

File: tmt/convert.py [hunk 26]
```diff
@@ -689,13 +725,13 @@ def read_nitrate_case(testcase, makefile_data=None, general=False):
...
                 testcase.tester.name, testcase.tester.email)
         else:
-            try:
-                # Use contact from Makefile if it's there and email matches
-                if re.search(testcase.tester.email, makefile_data['contact']):
-                    data['contact'] = makefile_data['contact']
-                else:
-                    raise KeyError
-            except (KeyError, TypeError):
+            if makefile_data is None:
```
Comment (thrix) [inline tmt/convert.py:728]: @ukulekek I think the original code here also handles if `makefile_data` dict does not contain `contact` key?

Comment (thrix) [reply tmt/convert.py:728]: maybe

```suggestion
            if makefile_data is None or 'contact' not in makefile_data:
```

?

File: tmt/convert.py [hunk 27]
```diff
@@ -770,7 +806,8 @@ def read_nitrate_case(testcase, makefile_data=None, general=False):
 
     # Extend bugs detected from Makefile with those linked in Nitrate
     try:
-        data['link'] = makefile_data['link'].copy()
+        if makefile_data is not None:
```
Comment (thrix) [inline tmt/convert.py:809]: if we extend the condition here, maybe we do not need the `try/except` clause?

```suggestion
        if makefile_data is not None and 'link' in makefile_data:
```

?

**Review by thrix — CHANGES_REQUESTED** [review]
@ukulekek found few more things after closer look

File: tmt/convert.py (hunk 24 above)
Comment (ukulekek) [reply tmt/convert.py:414]: Thanks, error is added

File: tmt/convert.py (hunk 27 above)
Comment (ukulekek) [reply tmt/convert.py]: Thanks, fixed

**Review by thrix — APPROVED** [review]
LGTM

**Comment by thrix** [comment]
Resolved conflict, waiting for tests and merging \o/

**Review by janhavlin — APPROVED** [review]

**Comment by psss** [comment]
> Resolved conflict, waiting for tests and merging \o/

Nice! Please, do a squash of the 7 commits before merging.

**Comment by thrix** [comment]
Amended commit message according to the standard. And finally can be merged.
